File: packages/ds-codegen/src/frameworks/swift/swiftui/radio-group.ts
import type { ComponentIR } from "../../../ir.js";
// Shared native emission-class substrate: the radio-collection facts are
// pure IR extraction (dom role, compositeControl, iteration, member
// bindings, defined-type alias) and live there once for both native
// targets (FEAT-COMPOSE-RADIO-ADMISSION-01).
import { radioGroupFacts } from "../../native-emission-class.js";

/** Native radio collection: identity, label, disabled state and writes all come from IR. */
export function emitNativeRadioGroup(ir: ComponentIR, exportName: string): string | undefined {
  const facts = radioGroupFacts(ir);
  if (!facts) return undefined;
  const {
    optionsProp,
    itemType: type,
    members,
    valueMember,
    labelMember,
    disabledMember,
    helpMember,
    disabledOptional,
    channel,
    defaultValueProp: defaultValue,
    labelProp,
    orientation,
  } = facts;
  const hasOrientation = orientation !== undefined;

  const params = [
    `${optionsProp}: [${type}]`,
    `${channel.valueProp}: Binding<String>? = nil`,
    ...(defaultValue ? [`${defaultValue}: String = ""`] : []),
    `${channel.changeHandlerProp}: ((String) -> Void)? = nil`,
    ...(labelProp ? [`${labelProp}: String = ""`] : []),
    ...(hasOrientation ? [`orientation: ${orientation.typeName} = .${orientation.defaultMember}`] : []),
  ];
  const lines = ["// @generated:start component", `public struct ${type} {`];
  for (const field of members) lines.push(`    public let ${field.name}: ${field.type}${field.optional ? "?" : ""}`);
  lines.push(`    public init(${members.map((field) => `${field.name}: ${field.type}${field.optional ? "? = nil" : ""}`).join(", ")}) {`);
  for (const field of members) lines.push(`        self.${field.name} = ${field.name}`);
  lines.push("    }", "}", "", "/// Native single-choice Picker; web form names have no SwiftUI transport meaning.",
    `public struct ${exportName}: View {`, `    private let ${optionsProp}: [${type}]`,
    "    @StateObject private var selection: ControllableValue<String>");
  if (labelProp) lines.push(`    private let ${labelProp}: String`);
  if (hasOrientation) lines.push(`    private let orientation: ${orientation.typeName}`);
  lines.push("", `    public init(${params.join(", ")}) {`, `        self.${optionsProp} = ${optionsProp}`,
    `        self._selection = StateObject(wrappedValue: ControllableValue(controlled: ${channel.valueProp}, defaultValue: ${defaultValue ?? '""'}, onChange: ${channel.changeHandlerProp}))`);
  if (labelProp) lines.push(`        self.${labelProp} = ${labelProp}`);
  if (hasOrientation) lines.push("        self.orientation = orientation");
  lines.push("    }", "", "    private var choices: some View {",
    `        SwiftUI.Picker(${labelProp ?? '""'}, selection: selection.binding()) {`,
    `            ForEach(${optionsProp}, id: \\.${valueMember}) { item in`,
    `                SwiftUI.Text(verbatim: item.${labelMember}).tag(item.${valueMember})`);
  if (disabledMember) lines.push(`                    .disabled(item.${disabledMember}${disabledOptional ? " ?? false" : ""})`);
  if (helpMember) {
    const optional = members.find((entry) => entry.name === helpMember)?.optional;
    lines.push(`                    .help(item.${helpMember}${optional ? ' ?? ""' : ""})`);
  }
  lines.push("            }", "        }", "        .pickerStyle(.radioGroup)", "    }", "", "    public var body: some View {");
  if (hasOrientation) lines.push("        if orientation == .horizontal {", "            choices.horizontalRadioGroupLayout()", "        } else {", "            choices", "        }");
  else lines.push("        choices");
  lines.push("    }", "}", "// @generated:end");
  return lines.join("\n");
}

File: packages/ds-swiftui/Sources/DsSwiftUI/Components/RadioGroup/RadioGroup.swift
// @generated:start imports
import SwiftUI
// @generated:end

// @generated:start types
public enum RadioGroupOrientation: String, CaseIterable {
    case vertical
    case horizontal
}
// @generated:end

// @generated:start component
public struct RadioGroupOption {
    public let value: String
    public let label: String
    public let disabled: Bool?
    public let description: String?
    public init(value: String, label: String, disabled: Bool? = nil, description: String? = nil) {
        self.value = value
        self.label = label
        self.disabled = disabled
        self.description = description
    }
}

/// Native single-choice Picker; web form names have no SwiftUI transport meaning.
public struct RadioGroup: View {
    private let options: [RadioGroupOption]
    @StateObject private var selection: ControllableValue<String>
    private let ariaLabel: String
    private let orientation: RadioGroupOrientation

    public init(options: [RadioGroupOption], value: Binding<String>? = nil, defaultValue: String = "", onChange: ((String) -> Void)? = nil, ariaLabel: String = "", orientation: RadioGroupOrientation = .vertical) {
        self.options = options
        self._selection = StateObject(wrappedValue: ControllableValue(controlled: value, defaultValue: defaultValue, onChange: onChange))
        self.ariaLabel = ariaLabel
        self.orientation = orientation
    }

    private var choices: some View {
        SwiftUI.Picker(ariaLabel, selection: selection.binding()) {
            ForEach(options, id: \.value) { item in
                SwiftUI.Text(verbatim: item.label).tag(item.value)
                    .disabled(item.disabled ?? false)
                    .help(item.description ?? "")
            }
        }
        .pickerStyle(.radioGroup)
    }

    public var body: some View {
        if orientation == .horizontal {
            choices.horizontalRadioGroupLayout()
        } else {
            choices
        }
    }
}
// @generated:end
